port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,413 @@
+ * instead of libc backtrace as it handles signal handler frames correctly, whereas backtrace may return an
+ * incorrect frame for the frame before the signal handler.
+ *
+ * @param[out] array the array to hold the frame addresses
+ * @param[in] size capacity of the array
+ *
+ * @return the number of addresses recorded in the array
+ */
+static uintptr_t
+unw_backtrace(void **array, const uintptr_t size)
+{
+	unw_cursor_t cursor;
+	unw_cursor_t prev;
+	unw_context_t uc;
+	unw_word_t ip = 0;
+	char buf[64];
+	uintptr_t i = 0;
+
+	unw_getcontext(&uc);
+	unw_init_local(&cursor, &uc);
+	while ((unw_step(&cursor) > 0) && (i < size)) {
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		array[i] = (void *)ip;
+		i += 1;
+
+		/* Save frame for check below */
+		memcpy(&prev, &cursor, sizeof(unw_cursor_t));
+	}
+
+	/* Darwin's libunwind cannot handle sigtramp in some cases. If the final frame iterated is sigtramp,
+	 * there are more frames but need to manually restore registers.
+	 */
+	if (0 == unw_get_proc_name(&prev, buf, sizeof(buf), &ip)) {
+		if (0 == strcmp(buf, "_sigtramp")) {
+			return handle_sigtramp(&prev, array, size, i);
+		}
+	}
+
+	return i;
+}
+
+/**
+ * NULL handler. We only care about preventing the signal from propagating up the call stack, no need to do
+ * anything in the handler.
+ */
+static uintptr_t
+handler(OMRPortLibrary *portLibrary, uint32_t gpType, void *gpInfo, void *userData)
+{
+	return OMRPORT_SIG_EXCEPTION_RETURN;
+}
+
+/**
+ * Wrapped call to libunwind backtrace functions
+ */
+static uintptr_t
+protectedBacktrace(OMRPortLibrary *port, void *arg)
+{
+	frameData *addresses = (frameData *)arg;
+	return unw_backtrace(addresses->address_array, addresses->capacity);
+}
+
+/**
+ * Provides signal protection for the unw_backtrace call. If the stack walk causes a segfault then we check the output
+ * array to see if we got any frames as we may be able to provide a partial backtrace. We also record that the stack walk
+ * was incomplete so that information is available after the function returns.
+ */
+static uintptr_t
+backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, const uintptr_t capacity)
+{
+	uintptr_t ret = 0;
+	frameData args;
+	args.address_array = address_array;
+	args.capacity = capacity;
+
+	memset(address_array, 0, sizeof(void *) * capacity);
+
+	if (NULL != omrthread_self()) {
+		if (0 != portLibrary->sig_protect(portLibrary, protectedBacktrace, &args, handler, NULL,
+				OMRPORT_SIG_FLAG_SIGALLSYNC | OMRPORT_SIG_FLAG_MAY_RETURN, &ret)
+		) {
+			/* check to see if there were any addresses populated */
+			for (ret = 0; (ret < args.capacity) && (NULL != address_array[ret]); ret++) {
+				/* increment ret to the last filled address */
+			}
+
+			threadInfo->error = FAULT_DURING_BACKTRACE;
+		}
+
+		return ret;
+	} else {
+		return unw_backtrace(address_array, capacity);
+	}
```

Comment:
if possible, multiple return statements should be avoided:
```suggestion
	} else {
		ret = unw_backtrace(address_array, capacity);
	}
	
	return ret;
```